crates/ty_project/src/db.rs

```diff
@@ -126,6 +138,59 @@ impl ProjectDatabase {
         )
     }
 
+    /// Deserialize the database from the persistent cache.
+    pub fn deserialize(&mut self, path: &str, metadata: &ProjectMetadata) -> anyhow::Result<()> {
+        // Read from the persistent cache.
+        let contents = match std::fs::read(path) {
+            Ok(contents) => contents,
+            Err(err) if err.kind() == io::ErrorKind::NotFound => return Ok(()),
+            Err(err) => return Err(err.into()),
+        };
+
+        let _guard = ValueSourceGuard::new(ValueSource::Cli, false);
```

Comment:
Does this mean that all deserialized options now have a source CLI even if they were originally loaded from a settings file?

Reply:
The only place I see the options being created is [with `ValueSource::Cli` here](https://github.com/astral-sh/ruff/blob/main/crates/ty/src/args.rs#L184). Are they created elsewhere with a different source?

Reply:
https://github.com/astral-sh/ruff/blob/f34b65b7a00dfc61c5a54e4280b421c94151e677/crates/ty_project/src/metadata.rs#L166-L177

Reply:
It only ignores it when serializing. The deserialization uses the serde_toml's spanned feature to get the span and we get the source from the thread local:

https://github.com/astral-sh/ruff/blob/f34b65b7a00dfc61c5a54e4280b421c94151e677/crates/ty_project/src/metadata/value.rs#L281-L308

We don't serialize the span and source because the range is only for better diagnostics. It isn't actually part of the data itself. Now, the good news is that we only use serialization for tests. So we would consider serializing the ranges (or only if a specific thread local is set) and deserialize the ranges from the data if present (over using the thread local). I guess, we could always serialize/deserialize the source if the thread local isn't set